### ISSUE TITLE
Cannot find module 'abstract-socket'

### ISSUE THREAD

**Issue by bhaugen** [ssbc/patchwork#382]
This morning I did
```$ npm install ssb-patchwork -g ```
And then
```
$ patchwork
error loading sodium bindings: Cannot find module 'chloridedown/build/Release/sodium'
falling back to javascript version.
[CFG] Password: NO.
[CFG] TLS: NO.
[CFG] Remote Access: NO.
Patchwork - Copyright (C) 2015-2016 Secure Scuttlebutt Consortium
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it under certain conditions (GPL-3.0).

Starting...
Loading plugins from /home/bob/.ssb/node_modules
Log level: notice
/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/libsodium/dist/modules/libsodium.js:14
var Module;if(!Module)Module=(typeof Module!=="undefined"?Module:null)||{};var moduleOverrides={};for(var key in Module){if(Module.hasOwnProperty(key)){moduleOverrides[key]=Module[key]}}var ENVIRONMENT_IS_WEB=typeof window==="object";var ENVIRONMENT_IS_WORKER=typeof importScripts==="function";var ENVIRONMENT_IS_NODE=typeof process==="object"&&typeof require==="function"&&!ENVIRONMENT_IS_WEB&&!ENVIRONMENT_IS_WORKER;var ENVIRONMENT_IS_SHELL=!ENVIRONMENT_IS_WEB&&!ENVIRONMENT_IS_NODE&&!ENVIRONMENT_IS_WORKER;if(ENVIRONMENT_IS_NODE){if(!Module["print"])Module["print"]=function print(x){process["stdout"].write(x+"\n")};if(!Module["printErr"])Module["printErr"]=function printErr(x){process["stderr"].write(x+"\n")};var nodeFS=require("fs");var nodePath=require("path");Module["read"]=function read(filename,binary){filename=nodePath["normalize"](filename);var ret=nodeFS["readFileSync"](filename);if(!r

Error: Cannot find module 'abstract-socket'
    at Function.Module._resolveFilename (module.js:326:15)
    at Function.Module._load (module.js:277:25)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at createStream (/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/dbus-native/index.js:45:19)
    at createConnection (/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/dbus-native/index.js:70:30)
    at Function.module.exports.createClient (/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/dbus-native/index.js:130:20)
    at Function.module.exports.sessionBus (/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/dbus-native/index.js:141:25)
    at EventEmitter.init (/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/freedesktop-notifications/lib/notifications.js:67:20)
    at init (/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/ssb-notifier/notifier.js:7:28)
```

**Comment by bhaugen** [comment]
Result of this bug: I am shut out of patchwork at present. Could probably use one of the dists but I sorta liked the npm installs. Any fixes on the horizon?

**Comment by pfraze** [comment]
@dominictarr you have any idea what could be causing this?

@bhaugen did you try reinstalling? Uninstalling and then reinstalling? There really shouldnt be anything causing this.

**Comment by bhaugen** [comment]
Tried reinstalling. Then uninstalled-reinstalled. Same result.

Using node 5.4.0 if that matters.

Got these warnings from uninstall, if they matter:
```
bob@ganoderma:~$ npm uninstall ssb-patchwork
npm WARN ENOENT ENOENT: no such file or directory, open '/home/bob/package.json'
npm WARN EPACKAGEJSON bob No description
npm WARN EPACKAGEJSON bob No repository field.
npm WARN EPACKAGEJSON bob No README data
npm WARN EPACKAGEJSON bob No license field.
```

**Comment by pfraze** [comment]
Add a `-g` to the uninstall. Not sure if that will make a difference, but it might.

**Comment by bhaugen** [comment]
Made the uninstall work better, but the install results are the same.

**Comment by pfraze** [comment]
It looks like dbus-native is the one using abstract-socket, and freedesktop-notifications is the module using dbus-native...

but dbus-native has its package.json correctly set up (it lists abstract-socket in there) so I'm now suspicious that npm is fucking up again. What version of npm are you using?

**Comment by bhaugen** [comment]
nvm use 5.4.0
npm@3.3.12

**Comment by pfraze** [comment]
yep, that's a broken version of npm. You need to upgrade it to 3.5, at least.

Go to `~/.nvm/versions/node/v5.4.0/lib` and run `npm update` and see if that does it for you. If not, try `npm install npm`.

**Comment by bhaugen** [comment]
Now on npm@3.9.0
npm install ssb-patchwork -g
same result.
Also got this new warning:
npm WARN optional Skipping failed optional dependency /ssb-patchwork/chokidar/fsevents:
npm WARN notsup Not compatible with your operating system or architecture: fsevents@1.0.12

I'll try the uninstall-reinstall later, out of time now...

**Comment by pfraze** [comment]
the fsevents thing might be related

also try putting the `-g` before the module name

**Comment by bhaugen** [comment]
Nope. And I did the whole uninstall-reinstall again.

**Comment by pfraze** [comment]
What happens if you go into `/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork` and run `npm install abstract-sockets`?

**Comment by bhaugen** [comment]
```
$ npm install abstract-sockets
npm ERR! Linux 3.2.0-101-generic
npm ERR! argv "/home/bob/.nvm/versions/node/v5.4.0/bin/node" "/home/bob/.nvm/versions/node/v5.4.0/bin/npm" "install" "abstract-sockets"
npm ERR! node v5.4.0
npm ERR! npm  v3.9.0
npm ERR! code E404

npm ERR! 404 Registry returned 404 for GET on https://registry.npmjs.org/abstract-sockets
npm ERR! 404 
npm ERR! 404  'abstract-sockets' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404 
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.

npm ERR! Please include the following file with any support request:
npm ERR!     /home/bob/npm-debug.log
```

**Comment by pfraze** [comment]
my bad, abstract-socket. It's not plural

**Comment by bhaugen** [comment]
Got a lot farther, but:

```
Writing output to GYP file deps/libsodium.gyp
ssb-patchwork@2.8.1 /home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork
└── abstract-socket@1.1.0  extraneous

npm WARN optional Skipping failed optional dependency /chokidar/fsevents:
npm WARN notsup Not compatible with your operating system or architecture: fsevents@1.0.12
npm ERR! Linux 3.2.0-102-generic
npm ERR! argv "/home/bob/.nvm/versions/node/v5.4.0/bin/node" "/home/bob/.nvm/versions/node/v5.4.0/bin/npm" "install" "abstract-socket"
npm ERR! node v5.4.0
npm ERR! npm  v3.9.0
npm ERR! path /home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/.staging/ssb-patchwork-0aae447a
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall rename

npm ERR! enoent ENOENT: no such file or directory, rename '/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/.staging/ssb-patchwork-0aae447a' -> '/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork'
npm ERR! enoent ENOENT: no such file or directory, rename '/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/.staging/ssb-patchwork-0aae447a' -> '/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent 
npm ERR! Linux 3.2.0-102-generic
npm ERR! argv "/home/bob/.nvm/versions/node/v5.4.0/bin/node" "/home/bob/.nvm/versions/node/v5.4.0/bin/npm" "install" "abstract-socket"
npm ERR! node v5.4.0
npm ERR! npm  v3.9.0
npm ERR! path npm-debug.log.1245843981
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall open

npm ERR! enoent ENOENT: no such file or directory, open 'npm-debug.log.1245843981'
npm ERR! enoent ENOENT: no such file or directory, open 'npm-debug.log.1245843981'
npm ERR! enoent This is most likely not a problem with npm itself
npm ERR! enoent and is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! Please include the following file with any support request:
npm ERR!     /home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/npm-debug.log
```

**Comment by pfraze** [comment]
I have no idea what this is. This is either because of `fsevents` not installing, or npm is screwing up.

**Comment by clehner** [comment]
I just published ssb-notifier@2.0.2 which should prevent errors like this from crashing sbot/patchwork:
```
...
    at createStream (/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/dbus-native/index.js:45:19)
    at createConnection (/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/dbus-native/index.js:70:30)
    at Function.module.exports.createClient (/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/dbus-native/index.js:130:20)
    at Function.module.exports.sessionBus (/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/dbus-native/index.js:141:25)
    at EventEmitter.init (/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/freedesktop-notifications/lib/notifications.js:67:20)
    at init (/home/bob/.nvm/versions/node/v5.4.0/lib/node_modules/ssb-patchwork/node_modules/ssb-notifier/notifier.js:7:28)
```
